File: compra.php
<?php

require_once 'vendor/autoload.php';

/**
 *
 * NubeAndo.COM - Educar es Compartir
 *
 * @author: Angel Céspedes Quiroz
 * @Usuario: acq1305
 * @Web: https://www.nubeando.com
 * @Facebook: https://www.facebook.com/acq1305
 * @Skype: acq1305
 * @Linkedin: https://bo.linkedin.com/in/acq1305
 *
 */

$app = new \Slim\Slim();

$db = new mysqli('localhost', 'root', '', 'farmaciacumbre');

// Configuración de cabeceras
header('Access-Control-Allow-Origin: *');
header("Access-Control-Allow-Headers: X-API-KEY, Origin, X-Requested-With, Content-Type, Accept, Access-Control-Request-Method");
header("Access-Control-Allow-Methods: GET, POST, OPTIONS, PUT, DELETE");
header("Allow: GET, POST, OPTIONS, PUT, DELETE");
header("Content-Type: application/json");

$method = $_SERVER['REQUEST_METHOD'];
if($method == "OPTIONS") {
    die();
}

/*$loginToken = function (){
    $app = \Slim\Slim::getInstance();
    $token = $app->request->headers('ApiKey');
    if($token=='1234567'){

    }else{
        throw new Exception("Usuario no valido");
    }
}*/

// LISTAR TODOS LOS PRODUCTOS
$app->get('/compras', function() use($db, $app){
    $sql = 'SELECT * FROM compra ORDER BY id DESC;';
    $query = $db->query($sql);

    $productos = array();
    while ($producto = $query->fetch_assoc()) {
        $productos[] = $producto;
    }

    $result = array(
        'status' => 'success',
        'code'	 => 200,
        'data' => $productos
    );

    echo json_encode($result);
});

// DEVOLVER UN SOLO PRODUCTO
$app->get('/compras/:id', function($id) use($db, $app){
    $sql = 'SELECT * FROM compra WHERE id = '.$id;
    $query = $db->query($sql);

    $result = array(
        'status' 	=> 'error',
        'code'		=> 404,
        'message' 	=> 'Producto no disponible'
    );

    if($query->num_rows == 1){
        $producto = $query->fetch_assoc();

        $result = array(
            'status' 	=> 'success',
            'code'		=> 200,
            'data' 	=> $producto
        );
    }

    echo json_encode($result);
});

// GUARDAR PRODUCTOS
$app->post('/compras', function() use($app, $db){
    $result = array(
        'status' => 'error',
        'code'	 => 404,
        'message' => 'Producto NO se ha creado'
    );

    // $token = $app->request->headers('ApiKey');

    // if($token=='1234567'){

        $json = $app->request->getBody('json');
        $data = json_decode($json, true);

        if(!isset($data['fechacompra'])){
            $data['fechacompra']=null;
        }

        if(!isset($data['nombre'])){
            $data['nombre']=null;
        }

        if(!isset($data['existencia'])){
            $data['existencia']=null;
        }

        if(!isset($data['cantidad'])){
            $data['cantidad']=null;
        }

        if(!isset($data['precio'])){
            $data['precio']=null;
        }

        if(!isset($data['idproveedor'])){
            $data['idproveedor']=null;
        }

        if(!isset($data['usuario_id'])){
            $data['usuario_id']=null;
        }

        $query = "INSERT INTO compra VALUES(NULL,".
            "'{$data['fechacompra']}',".
            "'{$data['nombre']}',".
            "'{$data['existencia']}',".
            "'{$data['cantidad']}',".
            "'{$data['precio']}',".
            "'{$data['idproveedor']}',".
            "'{$data['usuario_id']}'".
            ");";

        $insert = $db->query($query);

        if($insert){
            $result = array(
                'status' => 'success',
                'code'	 => 200,
                'message' => 'Producto creado correctamente'
            );
        }
    //}

    echo json_encode($result);

});

// ACTUALIZAR UN PRODUCTO
$app->put('/compras/:id', function($id) use($db, $app){
    $json = $app->request->getBody('json');
    $data = json_decode($json, true);

    $sql = "UPDATE compra SET ".
        "fechacompra = '{$data["fechacompra"]}', ".
        "nombre = '{$data["nombre"]}', ".
        "existencia = '{$data["existencia"]}', ".
        "cantidad = '{$data["cantidad"]}', ".
        "precio = '{$data["precio"]}', ".
        "idproveedor = '{$data["idproveedor"]}', ".
        "usuario_id = '{$data["usuario_id"]}' WHERE id = {$id} ";
       

   

    $query = $db->query($sql);

    if($query){
        $result = array(
            'status' 	=> 'success',
            'code'		=> 200,
            'message' 	=> 'El producto se ha actualizado correctamente!!'
        );
    }else{
        $result = array(
            'status' 	=> 'error',
            'code'		=> 404,
            'message' 	=> 'El producto no se ha actualizado!!'
        );
    }

    echo json_encode($result);

});

// ELIMINAR UN PRODUCTO
$app->delete('/compras/:id', function($id) use($db, $app){
    $sql = 'DELETE FROM compra WHERE id = '.$id;
    $query = $db->query($sql);

    if($query){
        $result = array(
            'status' 	=> 'success',
            'code'		=> 200,
            'message' 	=> 'El producto se ha eliminado correctamente!!'
        );
    }else{
        $result = array(
            'status' 	=> 'error',
            'code'		=> 404,
            'message' 	=> 'El producto no se ha eliminado!!'
        );
    }

    echo json_encode($result);
});

$app->run();

File: servicios/Compra.java
/**
 * @api {get} /compra/  GET COMPRA
 * @apiName Categoria 1
 * @apiGroup 4.- COMPRA_PRODUCTO
 *
 *
 * @apiSuccess {Number} Id Id Unico Compra Producto .
 * @apiSuccess {Date} Fecha_Compra  Registro de fecha Compra.
 * @apiSuccess {String} Nombre_Producto  Nombre Del Producto.
 * @apiSuccess {Number} Exitencia  Cuanto es la cantidad del Producto.
 * @apiSuccess {Number} Cantidad  Cantidad requeridad del producto .
 * @apiSuccess {Number} Precio  Precio del Producto.
 
* @apiSuccess {Number} IdProveedor  Llave foranea de Proveedor.
* @apiSuccess {Number} IdUsuario  Llave foranea de Usuario.

 * @api {get} /compra 1.- GET COMPRA PRODUCTO 
 * @apiSuccessExample {json} Success-Response:
 *     HTTP/1.1 200 OK
 *      {
 *           "status": "success",
 *           "code": "200"
 *       }
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 404 Not Found
 *    {
 *      'status' 	=> 'error',
 *      'code'		=> 404,
 *      'message' 	=> 'Compra no disponible'
 *     }
 */

 /**
 * @api {get} /compra/:id  GET compra Por ID
 * @apiName Categoria 2
 * @apiGroup 4.- COMPRA_PRODUCTO
 *
 * @apiParam {Number} Id Id Unico del Compra Producto .
 
 * @apiSuccess {Number} Id Id Unico Compra Producto .
 * @apiSuccess {Date} Fecha_Compra  Registro de fecha Compra.
 * @apiSuccess {String} Nombre_Producto  Nombre Del Producto.
 * @apiSuccess {Number} Exitencia  Cuanto es la cantidad del Producto.
 * @apiSuccess {Number} Cantidad  Cantidad requeridad del producto .
 * @apiSuccess {Number} Precio  Precio del Producto.
 
* @apiSuccess {Number} IdProveedor  Llave foranea de Proveedor.
* @apiSuccess {Number} IdUsuario  Llave foranea de Usuario.
 *
 * @api {get} /compras/:id 2.- GET COMPRA POR ID
 * @apiSuccessExample {json} Success-Response:
 *     HTTP/1.1 200 OK
 *       {
 *           "status": "success",
 *           "code": "200"
 *       }
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 404 Not Found
 *     {
 *      'status' 	=> 'error',
 *      'code'		=> 404,
 *      'message' 	=> 'Compra no disponible'
 *     }
 */

/**
 * @api {post} /categoria/ POST INSERTAR COMPRA

 * @apiName Categoria 3
 * @apiGroup 4.- COMPRA_PRODUCTO
 
 * @apiSuccess {Number} Id Id Unico Compra Producto .
 * @apiSuccess {Date} Fecha_Compra  Registro de fecha Compra.
 * @apiSuccess {String} Nombre_Producto  Nombre Del Producto.
 * @apiSuccess {Number} Exitencia  Cuanto es la cantidad del Producto.
 * @apiSuccess {Number} Cantidad  Cantidad requeridad del producto .
 * @apiSuccess {Number} Precio  Precio del Producto.
 
* @apiSuccess {Number} IdProveedor  Llave foranea de Proveedor.
* @apiSuccess {Number} IdUsuario  Llave foranea de Usuario.

 *
  * @api {post} /compras 3.-POST INSERTAR COMPRA
 * @apiSuccessExample {json} Success-Response:
 *     HTTP/1.1 200 OK
 *      {
 *      'status' => 'success',
 *      'code'	 => 200,
 *      'message' => 'Compra creada correctamente'
 *       }
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 404 Not Found
 *     {
 *      'status' => 'error',
 *      'code'	 => 404,
 *      'message' => 'Compra NO se ha creado'
 *     }
 */

 /**
 * @api {put} /compras/:id  PUT ACTUALIZAR COMPRA POR ID
 * @apiName Categoria 4
 * @apiGroup 4.- COMPRA_PRODUCTO
 *
 * @apiParam {Number} Id Id Unico del Compra Producto .
 
 * @apiSuccess {Number} Id Id Unico Compra Producto .
 * @apiSuccess {Date} Fecha_Compra  Registro de fecha Compra.
 * @apiSuccess {String} Nombre_Producto  Nombre Del Producto.
 * @apiSuccess {Number} Exitencia  Cuanto es la cantidad del Producto.
 * @apiSuccess {Number} Cantidad  Cantidad requeridad del producto .
 * @apiSuccess {Number} Precio  Precio del Producto.
 
 * @apiSuccess {Number} IdProveedor  Llave foranea de Proveedor.
 * @apiSuccess {Number} IdUsuario  Llave foranea de Usuario.
 *
 * @api {put} /compras/:id  4.- PUT ACTUALIZAR COMPRAS POR ID
 * @apiSuccessExample {json} Success-Response:
 *     HTTP/1.1 200 OK
 *      {
 *       'status' 	=> 'success',
 *       'code'		=> 200,
 *       'message' 	=> 'Se ha actualizado correctamente!!'
 *       }
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 404 Not Found
 *     {
 *       'status' 	=> 'error',
 *       'code'		=> 404,
 *       'message' 	=> 'Compra no se ha actualizado!!'
 *     }
 */


  /**
 * @api {delete} /compras/:id  DELETE ELIMINAR COMPRAS POR ID
 * @apiName Categoria 5
 * @apiGroup 4.- COMPRA_PRODUCTO
 *
 * @apiParam {Number} id ID unico Compra.
 *
 * @api {delete} /compras/:id  5.- DELETE ELIMINAR COMPRAS POR ID
 * @apiSuccessExample {json} Success-Response:
 *     HTTP/1.1 200 OK
 *      {
 *       'status' 	=> 'success',
 *       'code'		=> 200,
 *       'message' 	=> 'La Compra se ha eliminado correctamente!!'
 *       }
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 404 Not Found
 *     {
 *       'status' 	=> 'error',
 *       'code'		=> 404,
 *       'message' 	=> 'La Compra no se ha eliminado!!'
 *     }
 */